zanata-war/src/main/java/org/zanata/service/impl/ActivityServiceImpl.java

```diff
@@ -184,24 +185,34 @@ public Object getEntity(EntityType entityType, long entityId) {
     public void logTextFlowStateUpdate(@Observes(during = TransactionPhase.AFTER_SUCCESS) TextFlowTargetStateEvent event_) {
         // workaround for https://issues.jboss.org/browse/WELD-2019
         final TextFlowTargetStateEvent event = event_;
-        Long actorId = event.getActorId();
+
+        Long actorId = event.getKey().getActorId();
+
         if (actorId != null) {
             Lock lock = activityLockManager.getLock(actorId);
             lock.lock();
             try {
-                transactionUtil.run(() -> {
-                    HTextFlowTarget target =
-                            textFlowTargetDAO.findById(event.getTextFlowTargetId(),
-                                    false);
-                    HDocument document = documentDAO.getById(event.getDocumentId());
-                    ActivityType activityType =
-                            event.getNewState().isReviewed() ? ActivityType.REVIEWED_TRANSLATION
-                                    : ActivityType.UPDATE_TRANSLATION;
-
-                    logActivityAlreadyLocked(actorId,
-                            document.getProjectIteration(), target, activityType,
-                            target.getTextFlow().getWordCount().intValue());
-                });
+                HDocument document =
+                    documentDAO.getById(event.getKey().getDocumentId());
+
+                for (TextFlowTargetStateEvent.TextFlowTargetState state : event
+                    .getStates()) {
+                    transactionUtil.run(() -> {
+                        HTextFlowTarget target =
+                            textFlowTargetDAO.findById(
+                                state.getTextFlowTargetId(), false);
+                        ActivityType activityType =
+                            state.getNewState().isReviewed()
+                                ? ActivityType.REVIEWED_TRANSLATION
+                                : ActivityType.UPDATE_TRANSLATION;
+
+                        logActivityAlreadyLocked(actorId,
```

Comment:
@aeng The TFT is different each time, but ultimately we only store one Activity with (a) the total number of words translated and (b) a pointer to the last TFT. Don't we? That was the intention. 

If we're really generating one Activity for every TextFlowTarget (rolling up only when the TFT is modified within the hour), it's no wonder Activity is such a performance problem for us. How do all these Activities appear in the dashboard, eg when a translator translates 20 TFTs in the same document in an hour? Is that shown as 20 lines in the dashboard?